src/test/kotlin/org/eclipse/hara/ddiclient/integrationtest/HawkbitTimeForcedTest.kt

```diff
@@ -0,0 +1,319 @@
+            TestUtils.hawkbitUrl,
+            TestUtils.gatewayToken)
+
+        val client = HaraClientDefaultImpl()
+
+        val eventListener = object : MessageListener {
+            override fun onMessage(message: MessageListener.Message) {
+                "Message received: $message".internalLog()
+                when (message) {
+
+                    is MessageListener.Message.Event.UpdateFinished,
+                    MessageListener.Message.State.CancellingUpdate -> {
+                        testScope.launch {
+                            try {
+                                assertServerActionsOnCompleteJob?.await()
+                            } catch (ignored: CancellationException) {
+                            }
+                        }
+                    }
+
+                    else -> {
+                    }
+                }
+            }
+        }
+
+        client.init(
+            clientData,
+            directoryDataProvider,
+            configDataProvider,
+            deploymentPermitProvider,
+            listOf(eventListener, *messageListeners.toTypedArray()),
+            listOf(updater),
+            downloadBehavior,
+            httpBuilder = OkHttpClient.Builder().addOkhttpLogger()
+        )
+        client
+    }
+
+    @Test(enabled = true, timeOut = 150_000)
+    fun testTimeForcedUpdateWhileWaitingForDownloadAuthorization() = runBlocking {
+        reCreateTestTargetOnServer()
+
+        assignDistributionToTheTarget()
+
+        val client = createHaraClientWithPermissions(downloadAllowed = false)
+
+        val deployment = createTargetTestDeployment(testingForUpdateAuthorization = false)
+
+        startTheTestAndWaitForResult(client, deployment)
+
+    }
+
+    @Test(enabled = true, timeOut = 150_000)
+    fun testTimeForcedUpdateWhileWaitingForUpdateAuthorization() = runBlocking {
+        reCreateTestTargetOnServer()
+
+        assignDistributionToTheTarget()
+
+        val client = createHaraClientWithPermissions(downloadAllowed = true)
+
+        val deployment = createTargetTestDeployment(testingForUpdateAuthorization = true)
+
+        startTheTestAndWaitForResult(client, deployment)
+    }
+
+    private fun createHaraClientWithPermissions(
+        downloadAllowed: Boolean = false): HaraClient {
+
+        val deploymentBehavior = object : DeploymentPermitProvider {
+            override fun downloadAllowed() = CompletableDeferred(downloadAllowed)
+            override fun updateAllowed() = CompletableDeferred(false)
+        }
+        return defaultClientFromTargetId(
+            deploymentPermitProvider = deploymentBehavior).invoke(TARGET_ID)
+    }
+
+    private suspend fun startTheTestAndWaitForResult(client: HaraClient,
+                                                     deployment: TestUtils.TargetDeployments) {
+        client.startAsync()
+        assertServerActionsOnCompleteJob = assertServerActionsOnComplete(client, deployment)
+
+        testScope.async {
```

Comment:
Isn't better a launch/join instead of an async/await here?

Reply:
One distinction between launch/join and async/await lies in how they handle exceptions. 
Given the usage of multiple nested Coroutines in this test, it's essential that the inner Coroutine passes the caught exception to the main thread (Not just cancelling it's parent coroutine) to ensure that the test can appropriately fail.
For instance, using launch/join at this specific line may prevent the assertion exception from reaching the main thread, potentially causing the test to pass incorrectly.

https://stackoverflow.com/questions/53303358/kotlin-async-exception-handling
https://proandroiddev.com/kotlin-coroutines-patterns-anti-patterns-f9d12984c68e

Reply:
I understand now, thanks for the clarification.